tracing/contrib/aws/handlers.go

```diff
@@ -0,0 +1,139 @@
+package aws
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+var (
+	// Starts a span and adds it to the request context.
+	StartHandler = request.NamedHandler{
+		Name: "opentracing.Start",
+		Fn: func(r *request.Request) {
+			_, ctx := opentracing.StartSpanFromContext(r.Context(), "client.request")
+			r.SetContext(ctx)
+		},
+	}
+
+	// Adds information about the request to the span.
+	RequestInfoHandler = request.NamedHandler{
+		Name: "opentracing.RequestInfo",
+		Fn: func(r *request.Request) {
+			span := opentracing.SpanFromContext(r.Context())
+			span.SetTag("service.name", fmt.Sprintf("aws.%s", r.ClientInfo.ServiceName))
+			span.SetTag("resource.name", r.Operation.Name)
+			span.SetTag("http.method", r.Operation.HTTPMethod)
+			span.SetTag("http.url", r.ClientInfo.Endpoint+r.Operation.HTTPPath)
+			span.SetTag("out.host", r.ClientInfo.Endpoint)
+			span.SetTag("aws.operation", r.Operation.Name)
+		},
+	}
+
+	// Finishes the span.
+	FinishHandler = request.NamedHandler{
+		Name: "opentracing.Finish",
+		Fn: func(r *request.Request) {
+			span := opentracing.SpanFromContext(r.Context())
+			span.SetTag("aws.retry_count", fmt.Sprintf("%d", r.RetryCount))
+
+			if r.HTTPResponse != nil {
+				span.SetTag("http.status_code", fmt.Sprintf("%d", r.HTTPResponse.StatusCode))
+			}
+
+			if r.Error != nil {
+				span.SetTag("error.error", r.Error)
+				if err, ok := r.Error.(awserr.Error); ok {
+					span.SetTag("aws.err.code", fmt.Sprintf("%s", err.Code()))
+				}
+			}
+
+			span.Finish()
+		},
+	}
+
+	// DynamoDBInfoHandler adds extra information to the span for requests
+	// to DynamoDB.
+	DynamoDBInfoHandler = request.NamedHandler{
+		Name: "opentracing.DynamoDB",
+		Fn: func(r *request.Request) {
+			// Check if this request is for the DynamoDB API.
+			if r.ClientInfo.ServiceName != dynamodb.ServiceName {
+				return
+			}
+
+			tableName := dynamoDBTableName(r)
+			if tableName == nil {
+				return
+			}
+
+			span := opentracing.SpanFromContext(r.Context())
+			span.SetOperationName(r.Operation.Name)
+			span.SetTag("span.type", "db")
+			span.SetTag("resource.name", *tableName)
```

Comment:
Resource name is what we're able to monitor and alert on. This should be more specific than table name I think.

Reply:
Operation name plus table name would do the job I think.

Reply:
I think you'd still get that, since we're changing the operation name from `client.request` to `r.Operation.Name`. In DataDog, you'd get a metrics like `trace.GetItem`, `trace.PutItem`, etc, broken out by table name.

Reply:
Huh. Ok then. Works for me.